### PR TITLE
[BEAM-96] Add composed combine functions builders in CombineFns

### PR DESCRIPTION
**Pull request by peihe**
* compose() or composeKeyed() are used to start composition
* with() is used to add an input-transformation, a combine fn and an output TupleTag
* A non-CombineFn builder is used to ensure that every composition includes at least one item
* Duplicate output tags are not allowed in the same composition

### CODE REVIEW

**Comment by peihe** [comment]
R: @bjchambers

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 1]
```diff
@@ -0,0 +1,1071 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Static utility methods that create combine function instances.
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Static methods for composing {@link CombineFn CombineFns}.

Possibly elaborate.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 2]
```diff
@@ -0,0 +1,1071 @@
...
+          .with(extractInputFn, combineFnWithContext, outputTag);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * A tuple of {@link Object Objects} produced for a given key from the composed combine functions.
+   */
+  public static class CoCombineResult implements Serializable {
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Should this be a top-level class? This seems unrelated to the CombineFns namespace beyond the "it is the result of composing CombineFns"

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 3]
```diff
@@ -0,0 +1,1071 @@
...
+        TupleTag<OutputT> outputTag) {
+      return new ComposedCombineFnWithContext<DataT>()
+          .with(extractInputFn, combineFnWithContext, outputTag);
+    }
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * A tuple of {@link Object Objects} produced for a given key from the composed combine functions.
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: I'd avoid saying this is Objects. Conceptually, this *should* be a tuple of values, and the type of each value corresponds to what is output from the ComibneFn.

"A tuple of outputs produced by a composed {@link CombineFn}. See {@link #compose()} or {@link #composeKeyd()}) for details."

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 4]
```diff
@@ -0,0 +1,1071 @@
...
+
+    /**
+     * The constructor of {@link CoCombineResult}.
+     *
+     * @throws NullPointerException if any key or value in {@code mavaluesMapp} is null
+     */
+    CoCombineResult(
+        Map<TupleTag<?>, Object> valuesMap, Iterable<TupleTag<?>> tupleTags) {
+      this.valuesMap = ImmutableMap.copyOf(valuesMap);
+      this.tupleTags = ImmutableSet.copyOf(tupleTags);
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Why do we need tupleTags separately? We can use keySet() and containsKey, I believe, unless there are cases where tupleTags != valuesMap.keySet(). If there are, document them.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 5]
```diff
@@ -0,0 +1,1071 @@
...
+
+    @Override
+    public Object[] mergeAccumulators(Iterable<Object[]> accumulators) {
+      Iterator<Object[]> iter = accumulators.iterator();
+      if (!iter.hasNext()) {
+        return createAccumulator();
+      } else {
+        Object[] accum = iter.next();
+        for (int i = 0; i < combineFnCount; ++i) {
+          accum[i] = combineFns.get(i).mergeAccumulators(new ProjectionIterable(accumulators, i));
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: I assume that ProjectIterable(accumulators, i) will skip the first item in accumulators since we've already pulled that out? If yes, please document. If no, that seems wrong.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 1 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: We have four java combine function classes. I am using "combine function" to refer to all of them.
CombineFn is one of the class, and I think it will mislead some users.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 3 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Done.
But, replaced {@link CombineFn} with "combine functions". Same reason as above.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 4 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: They are not always the same. This is because I am using ImmutableMap, and it doesn't allow null values.
Documented

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 5 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: added comments:
        // Reuses the first accumulator, and modifies its values.
        // It is safe because {@code accum[i]} only depends on
        // the i-th component of each accumulator.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 2 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: If we have a combine sub-package, I would put CoCombineResult as a top-level class there.
But, we don't have that now, and I don't find other better locations.

For example, I don't like to put it as a top-level class in transforms, since it is not the data type from any Transforms.
And, I think it fit better in here, because CoCombineResult is the output of combine functions. And, it ties with how composed combine functions are implemented.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java [hunk 6]
```diff
@@ -0,0 +1,1071 @@
...
+   * A tuple of {@link Object Objects} produced for a given key from the composed combine functions.
+   */
+  public static class CoCombineResult implements Serializable {
+    private final Map<TupleTag<?>, Object> valuesMap;
+    private final Set<TupleTag<?>> tupleTags;
+
+    /**
+     * The constructor of {@link CoCombineResult}.
+     *
+     * @throws NullPointerException if any key or value in {@code mavaluesMapp} is null
```
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: mavaluesMapp -> valuesMap

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 4 above)
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: I think ImmutableSet has the same problem, so I'm guessing that the problem is with the value? It may be clearer to have allocate a static "NULL_VALUE" Object, and use that as a sentinel so you can put a non-nvull value in, and then use the single valuesMap everywhere.

If not, I think making it really explicit (`// Store tupleTags separately since tupleTags with null values are omitted`) would be good.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 2 above)
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Sounds fair. We should consider adding a top level "combine" package. One point of comparison, is that CoGroupByKey is in a "join" package, and CoGbkResult is top-level there as well.

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 5 above)
Comment (bjchambers) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Sounds great. Change "modifies its values" to "overwrites its values".

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 6 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Done

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 5 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Done

**Comment by peihe** [comment]
updated and pushed

File: sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java (hunk 4 above)
Comment (peihe) [inline sdk/src/main/java/com/google/cloud/dataflow/sdk/transforms/CombineFns.java]: Added NULL_VALUE, and tests
